### PR TITLE
fix: avoid stopping propagation when not necessary #11457

### LINKED ISSUE #11457: Tab component stopsPropagation on tab element click
## What happens?
When you attach your own js events to elements that are used as tab switchers, your events will not be fired.

## What should happen instead?
Your events should be executed normally.

## Possible Solution
https://github.com/zurb/foundation-sites/blob/develop/js/foundation.tabs.js#L162
Person who did this should be kicked in the nuts until they turn into an omelet, and then some more.
Also, remove all `.stopPropagation()` code from Foundation source. That smells.

## Context
I've spent 2 hours trying to figure out why my `$('.tab').on('click')` events are not firing. That's bloody ridiculous.

## Your Environment

- Foundation version(s) used: 6.2.4
- Browser(s) name and version(s): All JS capable browsers are affected
- Operating System and version (desktop or mobile): All JS capable browsers are affected

## Checklist (all required):

- [x] I have read and follow the CONTRIBUTING.md document.
- [x] There are no other issues similar to this one.
- [x] The issue title is descriptive.
- [x] The template is correctly filled.

### PR DESCRIPTION
**Pull request by ncoden**
## Description

Stopping the event propagation prevent the user to watch for the event at an higher level. If it was needed to prevent a duplicate behavior in a component, others methods should be used like debouncing the event.

This pull request remove `stopPropagation` and `stopImmediatePropagation` from everywhere we do not actually want to "cancel" the event.

### Changes:
- Save and check for `event.zfIsKeyHandled` in Keyboard utility to prevent to handle twice the same event in different components.
- Remove `stopPropagation` from components' keyboard handlers (Accordion, DropdownMenu, OffCanvas and Tabs).
- Remove `stopPropagation` and `stopImmediatePropagation` from everywhere we do not want to cancel the event (AccordionMenu, Drilldown, DropdownMenu, Tabs, Tooltip).
- Update `DropdownMenu._removeBodyHandler` to also ignore click events on the element itself.

### Others changes:
- Clean up `DropdownMenu._events` a bit.
- Add explainaitions where `stopPropagation` is still used (OffCanvas, Triggers).

### Actions taken to avoid regression:
- [x] Check for when `stopPropagation` was added, the reasons behind it and if it resolved any bugs (#8691).
- [x] Search for any related issues.
- [x] Run all visual tests.

### Related
- See [The Dangers of Stopping Event Propagation](https://css-tricks.com/dangers-stopping-event-propagation/) (@philipwalton)
- Closes #11457

## Types of changes

- [ ] Documentation
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
      functionality to change)

## Checklist (all required):

- [x] I have read and follow the [CONTRIBUTING](CONTRIBUTING.md) document.
- [x] There are no other pull request similar to this one.
- [x] The pull request title is descriptive.
- [x] The template is fully and correctly filled.
- [x] The pull request targets the right branch (`develop` or `develop-v...`).
- [x] My commits are correctly titled and contain all relevant information.
- [x] My code follows the code style of this project.
- ~~I have updated the documentation accordingly to my changes (if relevant).~~
- ~~I have added tests to cover my changes (if relevant).~~
- [x] All new and existing tests passed.

### CODE REVIEW

File: js/foundation.offcanvas.js [hunk 1]
```diff
@@ -296,13 +305,22 @@ class OffCanvas extends Plugin {
...
+   */
   _stopScrollPropagation(event) {
     let elem = this; // called from event handler context with this as elem
     let up = event.pageY < elem.lastY;
     let down = !up;
     elem.lastY = event.pageY;
 
     if((up && elem.allowUp) || (down && elem.allowDown)) {
+      // It is not recommended to stop event propagation (the user cannot watch it),
+      // but in this case this is the only solution we have.
```
Comment (SassNinja) [inline js/foundation.offcanvas.js]: well spotted 👍 
(`contentScroll: false` does not work correctly on iOS without this)

Comment (ncoden) [reply js/foundation.offcanvas.js]: Yes, I gone through the history of every `stopPropagation` and added comments when they were needed.

File: js/foundation.dropdownMenu.js [hunk 2]
```diff
@@ -286,8 +284,8 @@ class DropdownMenu extends Plugin {
     const $body = $(document.body);
     this._removeBodyHandler();
     $body.on('click.zf.dropdownMenu tap.zf.dropdownMenu', (e) => {
-      var $link = this.$element.find(e.target);
-      if ($link.length) return;
+      var isItself = this.$element.andSelf().find(e.target).length;
```
Comment (SassNinja) [inline js/foundation.dropdownMenu.js]: `andSelf()` is deprecated and should be replaced with `addBack()`

Apart from that I think the order is wrong. When you want to include the elem itself it should be
```javascript
var isItself = this.$element.find(e.target).addBack().length;
```

**Review by SassNinja — CHANGES_REQUESTED** [review]

File: js/foundation.dropdownMenu.js (hunk 2 above)
Comment (ncoden) [reply js/foundation.dropdownMenu.js]: I used `$(e.target).closest(this.$element).length`. I think that testing the DOM tree upward is way faster than browsing all the nodes downward.

Comment (SassNinja) [reply js/foundation.dropdownMenu.js]: Good idea 👍

**Review by SassNinja — APPROVED** [review]
LGTM

**Comment by ncoden** [comment]
Thank you for the review @SassNinja

**Comment by ncoden** [comment]
Targeted for `v6.6.0` because this may break things depending on the event listeners in the user project.

**Comment by joeworkman** [comment]
This pull request has been mentioned on **Foundation Open Source Community**. There might be relevant details there:

https://foundation.discourse.group/t/foundation-v6-6-1-v6-6-0-breaks-nested-dropdown-menus/1012/3

**Comment by DanielRuf** [comment]
@SassNinja there is a case where it breaks setups. Can you check https://foundation.discourse.group/t/foundation-v6-6-1-v6-6-0-breaks-nested-dropdown-menus/1012?

**Pull request merged**